src/libmongoc/src/mongoc/mongoc-topology-background-monitoring.c

```diff
@@ -0,0 +1,845 @@
+/*
+ * Copyright 2020-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc-topology-background-monitoring-private.h"
+
+#include "mongoc-client-private.h"
+#include "mongoc-log-private.h"
+#ifdef MONGOC_ENABLE_SSL
+#include "mongoc-ssl-private.h"
+#endif
+#include "mongoc-stream-private.h"
+#include "mongoc-topology-description-apm-private.h"
+#include "mongoc-topology-private.h"
+#include "mongoc-trace-private.h"
+#include "mongoc-util-private.h"
+
+#undef MONGOC_LOG_DOMAIN
+#define MONGOC_LOG_DOMAIN "bg_monitor"
+
+static BSON_THREAD_FUN (srv_polling_run, topology_void)
```

Comment:
Are there any tests for this method?

Reply:
Unfortunately, there are no automated tests. The SRV polling spec requires manual testing: https://github.com/mongodb/specifications/tree/master/source/polling-srv-records-for-mongos-discovery/tests which I ran through following these instructions: https://github.com/kevinAlbs/SRVPollingTesting. This did flush out some bugs. E.g. originally the polling function locked the topology mutex during the DNS call. This would block server selection and monitoring.

Unit testing this seemed out of scope for now. I think we'd need to somehow mock the DNS server, or introduce new test seams. But this could be additional motivation for introducing new testing infrastructure. I created https://jira.mongodb.org/browse/CDRIVER-3693 to track that.